pennylane/math/quantum.py

```diff
@@ -161,3 +162,103 @@ def marginal_prob(prob, axis):
     prob = np.reshape(prob, [2] * num_wires)
     prob = np.sum(prob, axis=inactive_wires)
     return np.flatten(prob)
+
+
+def _density_matrix_from_state_vector(state, wires, check_state=False):
+    """Compute the density matrix from a state vector.
+
+    Args:
+        state (tensor_like): 1D tensor state vector. This tensor should of size ``(2**N,)`` for some integer value ``N``.
+        wires (list(int)): List of wires (int) in the subsystem.

```

Comment:
@co9olguy We do not allow QNode returning state or density matrix to have different from the wires consecutive device, it is located in QubitDevice line 485

```
if self.wires.labels != tuple(range(self.num_wires)):
    raise qml.QuantumFunctionError(
        "Returning the state is not supported when using custom wire labels"
    )
```
We plan working on this soon.

Those functions needs consecutive indices because it can accept both state vectors and density matrices, from any framework and not necessarily linked to a QNode. It is more flexible in the argument but it requires to give consecutive wires. When the rework in the device class will be done, we just need to create a map before using the function

@albi3ro I have no problem to rename it indices and agree with your points 👍